lib/fog/orchestration/util/recursive_hot_file_loader.rb

```
@@ -0,0 +1,236 @@
+          #   with a suitable error.
+          #
+          # XXX Implement a retry here?
+          content = ''
+          # does ruby close the socket/filedescriptor at exit?
+          # XXX Limit download file size
+          # XXX Protect from vanilla open-uri attacks
+          uri_or_filename = uri_or_filename[7..-1] if uri_or_filename.start_with?("file:///")
+
+          open(uri_or_filename) { |f| content = f.read }
+          content
+        end
+
+        def get_template_contents(template_file)
+          # Retrieve a template content.
+          #
+          # @param template_file can be either:
+          #          - a raw_template string
+          #          - an URI string
+          #          - an Hash containing the parsed template.
+          #
+          # XXX: after deprecation of Ruby 1.9 we could use
+          #      named parameters and better mimic heatclient implementation.
+          Fog::Logger.warning("get_template_contents #{template_file}")
+
+          raise ArgumentError, "template_file should be Hash or String" unless
+            template_file.kind_of?(String) || template_file.kind_of?(Hash)
+
+          local_base_url = base_url_for_url(normalise_file_path_to_url(Dir.pwd + "/TEMPLATE"))
+
+          if template_file.kind_of?(Hash)
+            template_base_url = local_base_url
+            raw_template = YAML.dump(template_file)
+          elsif template_is_raw?(template_file)
+            raw_template = template_file
+            template_base_url = local_base_url
+          elsif template_is_url?(template_file)
+            template_file = normalise_file_path_to_url(template_file)
+            template_base_url = base_url_for_url(template_file)
+            raw_template = get_content(template_file)
+
+            Fog::Logger.warning("Template visited: #{@visited}")
+            @visited[template_file] = true
+          else
+            raise NotImplementedError, "template_file is not a string of the expected form"
```

Can I raise ArgumentError here too?

---

Neither is close to what you mean. We can simply raise a string error
```
  raise "template_file is not ..."
```